src/extra/gpkg/read.jl

```diff
@@ -0,0 +1,470 @@
+  flags::UInt8
+  srs_id::Int32
+  min_x::Float64
+  max_x::Float64
+  min_y::Float64
+  max_y::Float64
+  min_z::Union{Float64,Nothing}
+  max_z::Union{Float64,Nothing}
+  min_m::Union{Float64,Nothing}
+  max_m::Union{Float64,Nothing}
+end
+
+function parse_gpb_header(blob::Vector{UInt8})
+  io = IOBuffer(blob)
+  
+  magic = read(io, UInt16)
+  magic == 0x5047 || error("Invalid GeoPackage binary header")
+  
+  version = read(io, UInt8)
+  flags = read(io, UInt8)
+  
+  srs_id = ltoh(read(io, Int32))
+  
+  envelope_type = flags >> 1 & 0x07
+  has_envelope = envelope_type > 0
+  
+  if has_envelope && envelope_type >= 1
+    min_x = ltoh(read(io, Float64))
+    max_x = ltoh(read(io, Float64))
+    min_y = ltoh(read(io, Float64))
+    max_y = ltoh(read(io, Float64))
+  else
+    # Default envelope when not present
+    min_x = max_x = min_y = max_y = 0.0
+  end
+  
+  # Initialize z and m coordinates based on envelope type
+  if has_envelope && envelope_type >= 2
+    min_z = ltoh(read(io, Float64))
+    max_z = ltoh(read(io, Float64))
+  else
+    min_z = max_z = nothing
+  end
+  
+  if has_envelope && envelope_type >= 3
+    min_m = ltoh(read(io, Float64))
+    max_m = ltoh(read(io, Float64))
+  else
+    min_m = max_m = nothing
+  end
+  
+  header = GeoPackageBinaryHeader(
+    magic, version, flags, srs_id,
+    min_x, max_x, min_y, max_y,
+    min_z, max_z, min_m, max_m
+  )
+  
+  return header, position(io)
+end
+
+function read_coordinate_values(io::IOBuffer, byte_order::UInt8, has_z::Bool, has_m::Bool)
+  read_value = byte_order == 0x01 ? ltoh : ntoh
+  
+  x = read_value(read(io, Float64))
+  y = read_value(read(io, Float64))
+  
+  coords = if has_z
+    z = read_value(read(io, Float64))
+    (x, y, z)
+  else
+    (x, y)
+  end
+  
+  # skip M coordinate if present (we don't use it)
+  if has_m
+    read_value(read(io, Float64))
+  end
+  
+  return coords
+end
+
+function parse_wkb_geometry(blob::Vector{UInt8}, offset::Int, crs_type)
+  io = IOBuffer(blob[offset+1:end])
+  
+  byte_order = read(io, UInt8)
+  read_value = byte_order == 0x01 ? ltoh : ntoh
+  
+  wkb_type = read_value(read(io, UInt32))
+  
+  base_type = wkb_type & 0x0FFFFFFF
+  has_z = (wkb_type & WKB_Z) != 0
+  has_m = (wkb_type & WKB_M) != 0
+  
+  if base_type == WKB_POINT
+    return parse_wkb_point(io, byte_order, has_z, has_m, crs_type)
+  elseif base_type == WKB_LINESTRING
+    return parse_wkb_linestring(io, byte_order, has_z, has_m, crs_type)
+  elseif base_type == WKB_POLYGON
+    return parse_wkb_polygon(io, byte_order, has_z, has_m, crs_type)
+  elseif base_type == WKB_MULTIPOINT
+    return parse_wkb_multipoint(io, byte_order, has_z, has_m, crs_type)
+  elseif base_type == WKB_MULTILINESTRING
+    return parse_wkb_multilinestring(io, byte_order, has_z, has_m, crs_type)
+  elseif base_type == WKB_MULTIPOLYGON
+    return parse_wkb_multipolygon(io, byte_order, has_z, has_m, crs_type)
+  else
+    error("Unsupported WKB geometry type: $base_type")
+  end
```

Comment:
You can reduce the complexity of inner function calls by passing a single function with the recipe to read values from the buffer. Given `byte_order`, `has_z` and `has_m`, I believe you can define a function that takes `io` as input and consumes a single set of coordinates. Let's call this function `parse_coords`. You can then pass this function to all `parse_wkb_*` inner functions: `parse_wkb_point(io, parse_coords, crs_type)`.